### PR TITLE
Design doc - Publish Dandisets that contain Zarr archives

### PR DESCRIPTION
**Pull request by kabilar**
- [ ] Define user requirements
  - [x] Update design doc based on 2024-01-26 discussion @kabilar 
- [ ] Define technical specifications @waxlamp @kabilar 
- [ ] Prototype and evaluate tradeoffs of potential solutions
  - [ ] Pure DANDI side based on zarr manifest files
    - [x] Create manifest files with paths and version IDs for each chunk for a specific version of the Zarr archive @yarikoptic
       - [x] Implemented in this repo: https://github.com/dandi/zarr-manifests/  with more complete copy at https://datasets.datalad.org/?dir=/dandi/zarr-manifests/zarr-manifests-v2-sorted 
     - [x] Demonstrate use in dandidav: planned in https://github.com/dandi/dandidav/issues/43
    - [ ] Work out design for support in dandi-archive to adopt such an approach
  - [ ] Explore Arraylake @waxlamp

### CODE REVIEW

File: doc/design/zarr-publish-1.md [hunk 1]
```diff
@@ -30,7 +35,9 @@ The publishing procedure would follow the description found in the [publish-1 de
 
 ## MVP Technical specifications
 
-1. TODO
+1. Support versioning of Zarr archives.
+1. Create a unique web address for each published version of the Zarr archive.
```
[the remark this answers is not in the document]
Comment (kabilar) [reply doc/design/zarr-publish-1.md:39]: Thank you @yarikoptic.

File: .gitignore [hunk 2]
```diff
@@ -125,3 +125,5 @@ dmypy.json
 
 # Editor settings
 .vscode
+
+.DS_Store
```
Comment (waxlamp) [inline .gitignore:129]: Stuff like this, that's not directly related to a given PR, should really go in their own PRs (even if just a tiny one to update the .gitignore file).

**Review by waxlamp — COMMENTED** [review]
Leaving this review as a comment because it seems this is still WIP. I will also throw the switch to convert this PR to a draft.

File: doc/design/zarr-publish-1.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
+# Publish Dandisets that contain Zarr archives, and support updates to the Zarr archive after publishing the Dandiset
+
+This document describes the current implementation of publishing Dandisets with Zarr archives, desired use cases, and the associated requirements.
+
+## Current implementation
+
+When a blob asset is updated, a new version (i.e. a copy) is uploaded to the S3 bucket.  Zarr archives are too large so multiple copies should not be created.  A Zarr archive is uploaded once and it is updated in place.  This design means that the Zarr archive is immutable once the Dandiset is published, so that the published Dandiset is immutable. Currently, a Dandiset cannot be published if it contains a Zarr asset.  For more details, see the [zarr-support-3 design doc](https://github.com/dandi/dandi-archive/blob/master/doc/design/zarr-support-3.md).
```
[the remark this answers is not in the document]
Comment (satra) [reply doc/design/zarr-publish-1.md:7]: i would add that there are a two requirements.
- updating/publishing in a dandiset
- adding the asset to another dandiset 
- being able to publish both

for implementation:
- we currently don't have a readonly mode for an asset and any modification of zarr added to another dandiset should fail, unless control is handed over (more complicated).  
- adding should have either a link (readonly) or copy option. on the blob side, we have copy on write, but for assets this could get complicated for large trees.

File: .gitignore (hunk 2 above)
Comment (kabilar) [reply .gitignore]: Thank you.  I have reverted this commit and submitted it as a new pull request (#1863).

File: doc/design/zarr-publish-1.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+## Use case 2
+
+Upload a Zarr archive to an embargoed Dandiset. 
+
+## Use case 3
+
+Reuse a Zarr archive in more than one Dandiset.
+
+Allow for a Zarr archive that is uploaded as part of an original Dandiset to be packaged in a new Dandiset without duplicating the Zarr archive.  The new Dandiset could be created by potentially different authors and could contain additional raw and/or analyzed data.  This feature has been previously implemented for other asset types with [add_asset_to_dandiset.py](https://gist.github.com/satra/29404d965226e4c99fb48e7502953503#file-add_asset_to_dandiset-py).  Further details of this feature request have been previously documented in [dandi-archive #1792](https://github.com/dandi/dandi-archive/issues/1792).
```
[the remark this answers is not in the document]
Comment (kabilar) [reply doc/design/zarr-publish-1.md]: I agree that it is ideal to add support for this feature directly in the archive.

File: doc/design/zarr-publish-1.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+When a blob asset is updated, a new version (i.e. a copy) is uploaded to the S3 bucket.  Zarr archives are too large so multiple copies should not be created.  A Zarr archive is uploaded once and it is updated in place.  This design means that the Zarr archive is immutable once the Dandiset is published, so that the published Dandiset is immutable. Currently, a Dandiset cannot be published if it contains a Zarr asset.  For more details, see the [zarr-support-3 design doc](https://github.com/dandi/dandi-archive/blob/master/doc/design/zarr-support-3.md).
+
+## Use case 1
+
+Publish a Dandiset containing a Zarr archive(s), and subsequently update the Zarr archive(s).
+
+The publishing procedure would follow the description found in the [publish-1 design doc](https://github.com/dandi/dandi-archive/blob/master/doc/design/publish-1.md).  A modified publishing procedure that includes Zarr archive(s) is summarized below.
+
+1. User uploads a new Dandiset which includes a Zarr archive(s).
+2. User uploads an updated Zarr archive(s) to the `Draft` version of the Dandiset.
```
[the remark this answers is not in the document]
Comment (kabilar) [reply doc/design/zarr-publish-1.md]: I reordered steps 2 and 3, which would reflect functionality that is currently not in place (i.e. publishing a Dandiset with a Zarr archive and subsequently updating the `Draft` version).  Please let me know if I misunderstood your comment.

File: doc/design/zarr-publish-1.md [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+## MVP Technical specifications
+
+1. Support versioning of Zarr archives.
+1. Create a unique web address for each published version of the Zarr archive.
+1. Provide access to the Zarr archive versions through the web app and command line interface.
+1. TODO: add additional specifications
+
+## MVP+1 Technical specifications
+
+1. TODO
```
[the remark this answers is not in the document]
Comment (kabilar) [reply doc/design/zarr-publish-1.md:50]: Thanks Roni.  I have scheduled a meeting for tomorrow.  Let me know if its too short of notice and we can reschedule to next week.

**Comment by kabilar** [comment]
Closing this pull request as I am re-forking `dandi/dandi-archive` directly to `kabilar`.  Will reopen a new pull request.
